### ISSUE TITLE
Pass error msgs from HDF5-functions with `splash::DCExceptions`

### ISSUE THREAD

**Issue by BeyondEspresso** [ComputationalRadiationPhysics/libSplash#160]
Repeated crashes of PIConGPU during checkpoint writing (`splash::DCExceptions`:  Exception for DCDataSet [x] write: Failed to write dataset ) have risen the question, whether some bad input or a full file system (which according to `df -h .` was not supposed to be full) causes the exception.

Since libSplash only shows the fact that something went wrong (return value `<0`) but not the root cause, it would be helpful for debugging to extract the error stack with the HDF5 error handling functions and pass it along with the exception message in `stderr`.

**Comment by ax3l** [comment]
Did you try `export SPLASH_VERBOSE=<HIGH_NUMBER>` to get a bit more insights?

**Comment by BeyondEspresso** [comment]
No, I did not try `SPLASH_VERBOSE`. Instead I looked at the origin of the exceoption in the source code. There it is quite clear, that internal HDF5 error codes are pretty much ignored, since only the distinction between success and error is made.

I actually had a look into Chapter 9 and some other sections of the HDF5 manual. The best way foward seems to me to add an HDF5-error code to the interface of `splash::DCExceptions` and replace all occurances of this exception that handle failed HDF5-calls.

**Comment by bussmann** [comment]
I like the idea of forwarding HDF5 errors.

**Comment by BeyondEspresso** [comment]
It should be actually easier than described above. I only need to get the error stack in strings instead of having them shuffled into a file pointer.

**Comment by f-schmitt-zih** [comment]
I think you actually have to recompile libSplash with a special debug flag
to enable the hdf5 messages. The problem is that you will also get false
negatives from hdf5.

**Comment by BeyondEspresso** [comment]
I have a compiling branch, but I still need to test it somehow. Any idea how I can quickly provoke a nice hdf5-error without having to load the HDF5-error stack manually?

**Comment by f-schmitt-zih** [comment]
Compile libsplash with
-DSPLASH_VERBOSE_HDF5=1

**Comment by BeyondEspresso** [comment]
Current dev branch
`cmake -DCMAKE_INSTALL_PREFIX=~/lib/splash -DSPLASH_VERBOSE_HDF5=1 ~/src/libSplash`
```
Manually-specified variables were not used by the project:

    SPLASH_VERBOSE_HDF5
```

**Comment by ax3l** [comment]
@f-schmitt-zih suggested you should add such a flag if one wants to enable that.

just a reminder: exporting `SPLASH_VERBOSE` does not give you HDF5 errors but  it *still helps you* to check if your input is valid (since that was your initial question).

**Comment by ax3l** [comment]
@f-schmitt-zih we could alternatively add the HDF5 output just for a certain value of `SPLASH_VERBOSE` - or do you see an advantage to separate that?

**Comment by f-schmitt-zih** [comment]
No, I suggested to compile libsplash with that flag since it will enable
hdf5 error messages. The appropriate cmake option to use is DEBUG_VERBOSE.

As I said, enabling hdf5 error messages will also give you false negatives.
Don't know if they have an api or so to query them manually on request.

**Comment by ax3l** [comment]
Ah good to know! Why did the get a configure error then?

They have a manual API, I linked it above.

**Comment by BeyondEspresso** [comment]
This is strange. According to `CMakeLists.txt`, `src/SerialDataCollector.cpp` and `src/ParallelDataCollector.cpp`, an *activated* `DEBUG_VERBOSE` flag *supresses* all HDF5 output. A *deactivated* `DEBUG_VERBOSE` defaults to the library default settings, which is to output all HDF5 exceptions to `stderr`. However, I did not see any of these in the original `stderr` output.

Two questions:
1) Has the libsplash/1.2.2 module on hypnos been compiled with `DEBUG_VERBOSE` on or off?
2) Has direct output to `stderr` in libsplash any effect on the `stderr` in `picongpu`. (Is there anything else suppressed.)

**Comment by BeyondEspresso** [comment]
@f-schmitt-zih What do you mean with false negatives? Does this mean, the HDF5 API tells you everything is fine (negative, no error), when in reality it has thrown an exception?

**Comment by f-schmitt-zih** [comment]
You are wrong. DEBUG_VERBOSE sets SPLASH_VERBOSE_HDF5 which means the
ifndef is false and suppression us not triggered.

**Comment by f-schmitt-zih** [comment]
No, it shows an error when there is none.

**Comment by BeyondEspresso** [comment]
Thanks @f-schmitt-zih , somehow I have not seen the `n` in `ifndef`. If the `DEBUG_VERBOSE` option triggers false alarm, it would be useful to have splash append the HDF5 error stack to the message of the relevant `splash::DCException`s, so that there is only a report when a fatal error occurs. Then I will proceed with the pull request.
